packages/actor-query-operation-bgp-left-deep-reordering/lib/ActorQueryOperationBgpLeftDeepReordering.ts

```diff
@@ -0,0 +1,196 @@
+    const clusters = ActorQueryOperationBgpLeftDeepReordering.findConnectedPatterns(pattern.patterns);
+    // find cluster with least unbound variables (with the least patterns in case of a tie)
+    const clusterValues = clusters.map((cluster) => {
+      const val = -(pattern.patterns.length * cluster.variables.length + cluster.quads.length);
+      return { val, cluster };
+    });
+
+    clusterValues.sort((a, b) => {
+      return a.val - b.val;
+    });
+
+    const subPattern = clusterValues.pop().cluster.quads;
+
+    const best = {
+      count: Infinity,
+      index: -1 };
+
+    const outputs = (await Promise.all(subPattern.map(
+      (quad) => this.mediatorQueryOperation.mediate({ operation: quad, context }))))
+      .map(ActorQueryOperation.getSafeBindings);
+    const metadatas = await Promise.all(outputs.map((output) => output.metadata ? output.metadata() : <any> {}));
+
+    for (let i = 0; i < subPattern.length; ++i) {
+      const count = metadatas[i].totalItems;
+      if (count < best.count) {
+        if (best.index >= 0) {
+          // prevent useless nextPage calls
+          outputs[best.index].bindingsStream.close();
+        }
+        best.index = i;
+        best.count = count;
+      } else {
+        outputs[i].bindingsStream.close();
+      }
+    }
+    // can happen if there is not sufficient metadata
+    if (best.index < 0) {
+      best.index = 0;
+      outputs[0] = ActorQueryOperation.getSafeBindings(
+        await this.mediatorQueryOperation.mediate({ operation: subPattern[0], context }));
+    }
+
+    subPattern.splice(best.index, 1);
+
+    // create stream for the subPattern results
+    let bindingsStream = outputs[best.index].bindingsStream;
+    if (subPattern.length > 0) {
+      bindingsStream = new MultiTransformIterator<Bindings, Bindings>(bindingsStream);
+      (<MultiTransformIterator<Bindings, Bindings>> bindingsStream)._createTransformer = (bindings: Bindings) => {
+        const bgp = this.factory.createBgp(subPattern.map((quad) => this.bindPattern(quad, bindings)));
+        return this.getBoundOperationIterator(bgp, bindings, context);
+      };
+    }
+
+    // join with the remaining patterns
+    if (clusterValues.length > 0) {
+      bindingsStream = new MultiTransformIterator<Bindings, Bindings>(bindingsStream);
+      const patterns = Array.prototype.concat(...clusterValues.map((c) => c.cluster.quads));
+      const bgp = this.factory.createBgp(patterns);
+      (<MultiTransformIterator<Bindings, Bindings>> bindingsStream)._createTransformer = (bindings: Bindings) => {
+        return this.getBoundOperationIterator(bgp, bindings, context);
+      };
+    }
+
+    const variables = clusters
+      .reduce((vars: RDF.Term[], cluster) => vars.concat(
+        cluster.variables.filter((v) => vars.findIndex((val) => val.equals(v)) < 0)), [])
+      .map((term) => termToString(term));
+
+    return { type: 'bindings', bindingsStream, variables };
```

Comment:
Originally I didn't do it because it would require calling the metadata from all patterns not the cluster. But now that metadata is in a separate function I will add it. I do wonder how useful this value is going to be for anyone though. :D